MultiplexReservationSystem/src/main/java/p/lodz/pl/multiplexreservationsystem/service/ReservationsService.java

```diff
@@ -0,0 +1,139 @@
+        LocalDateTime expirationTime = calculateExpirationDate(screening);
+        reservation.setExpirationTime(expirationTime);
+
+        reservationsRepository.save(reservation);
+        bookTickets(reservation.getTickets());
+
+        return getReservationInfo(reservation);
+    }
+
+    public ReservationDto getReservationInfo(Reservations reservation) {
+        return mapToReservationsDto(reservation);
+    }
+
+    private void bookSeats(List<Tickets> tickets, Long screeningId) {
+        for (var ticket : tickets) {
+            bookedSeatRepository.save(new BookedSeat(ticket.getSeatId(), screeningId));
+        }
+    }
+
+    private void bookTickets(List<Tickets> tickets) {
+        ticketsRepository.saveAll(tickets);
+    }
+
+    private LocalDateTime calculateExpirationDate(Screenings screening) {
+        LocalTime movieDuration = moviesRepository.findById(screening.getMovieId()).orElseThrow().getDuration();
+        return screening.getDate().plusHours(movieDuration.getHour())
+                .plusMinutes(movieDuration.getMinute());
+    }
+
+    private boolean isTooLateToBook(Screenings screening) {
+        return LocalDateTime.now().isAfter(screening.getDate().minusMinutes(15));
+    }
+
+    private boolean isBookingCorrect(long screeningId, long roomId, List<Tickets> tickets) {
+        List<Seats> allSeats = seatsService.getRoomsSeats(roomId);
+        Set<Seats> seatsToBook = new HashSet<>(); //set for seats uniqueness
+        for (var ticket : tickets) {
+            seatsToBook.add(seatsService.getSeatById(ticket.getSeatId()));
+        }
+        List<Seats> bookedSeats = seatsService.getBookedSeats(roomId, screeningId);
+        List<Seats> availableSeats = seatsService.getAvailableSeats(roomId, bookedSeats);
+
+
+        if (seatsToBook.size() != tickets.size()
+                || !availableSeats.containsAll(seatsToBook)) {
+            return false;
+        }
+
+        for (var seat : seatsToBook) {
+            if (bookedSeats.contains(seat)
+                || !searchLeftNeighbours(allSeats, seatsToBook, bookedSeats, seat)
+                || !searchRightNeighbours(allSeats, seatsToBook, bookedSeats, seat)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean searchLeftNeighbours(List<Seats> allSeats, Set<Seats> seatsToBook,
+                                         List<Seats> bookedSeats, Seats seat) {
+        if (seat.getSeatNumber() != 1) {
+            Seats leftNeighbour = allSeats.stream().filter(s -> s.getId() == (seat.getId() - 1))
+                    .findFirst().orElseThrow();
+
+            if (!seatsToBook.contains(leftNeighbour)
+                    && !bookedSeats.contains(leftNeighbour)
+                    && leftNeighbour.getSeatNumber() != 1) {
+                Seats secondLeftNeighbour = allSeats.stream().filter(s -> s.getId() == (leftNeighbour.getId() - 1))
+                        .findFirst().orElseThrow();
```

Comment:
To nie poleci zawsze jak ktoś będzie rezerwował skrajne miejsce?

Reply:
Na początku obu metod sprawdzających lewych i prawych sąsiadów sprawdzam pierwszym if-em, czy wybrane miejsce nie jest skrajne. Jeśli jest np. lewe skrajne, to odrzucam sprawdzanie lewych sąsiadów i przechodze do sprawdzenia prawych.